### ISSUE TITLE
DnD Character - confused about the purpose and execution of the exercise

### ISSUE THREAD

**Issue by ireddick** [exercism/typescript#1057]
I'm unsure where to raise this, as having looked at some other tracks, what I found confusing is cross-track. However as I encountered it on this track I thought I'd raise it here.

I recently started working my way through the TypeScript track, and was making decent progress until I reached the DnD Character exercise. For context, I've used Exercism in the past, so am fairly familiar with the common early exercises, though I'd never tried DnD Character before now.

My question is: **what is the student learning here, and how are the tests helping them?**
ie. are we learning about randomness, or how to derive the DnD character from a series of dice rolls?

The exercise README describes the domain of generating a DnD character from a series of dice rolls, and lays out the calculations, with examples, for the basic abilities and the derived hitpoints.

The test suite, however:

  - Starts with a large number of tests that check the constitution modifier (though it's called 'ability modifier', which threw me for a while)

  - Has a large number of tests that check that the basic abilities are a within a range. This neither tests randomness, nor any kind of calculation - ie. the ability value from dice roll calculation described in the README.

  - Does test the hitpoints calculation, though with only one example and it depends on part of the implementation

  - The last test I don't understand at all regarding its purpose or what it's asserting:

    ``` javascript
    it('Each ability is only calculated once', () => {
      const character = new DnDCharacter()

      expect(character.strength === character.strength).toBeTruthy()
    })
    ```

Additionally, all tests are enabled to begin with instead of incrementally in a test first manner, meaning the student is in 'red' until they have a fully working implementation, or have to disable a lot of tests at the beginning.

In terms of where this exercise lies in the TypeScript track, students will have just finished 'Space Age'. What I'd be expecting here is maybe working with non-primitive input, a more complicated algorithm, and non-primitive output. I think what's described in the README *is* this, though not with the conflation with randomness that's in the test suite. I wouldn't expect to see testing randomness until much later in the track, and I'd likely not involve other rules like generating a DnD character in the same exercise.

I've looked at the common problem specification for this exercise, and while I think it is also a little confusing in talking about 'random', it mostly talks about property based tests for the calculations, which, as I understand things, isn't to do with testing randomness. I've also looked at this exercise in a couple of other tracks, and they share the same tests and test suite structure as the TypeScript version.

In summary, I think this exercise is pretty confusing as it is now (maybe enough to put someone off), and that the randomness ideas in the test suite aren't helpful re the problem domain in the README, or for learning about randomness. My suggestion is to change the exercise to be more about accepting a series of dice rolls and returning a character data structure that can be asserted against, which would be more in line with what's described in the README.

Apologies for this being so long! 🙂

**Comment by SleeplessByte** [comment]
I think these are very good questions, but the answers aren't exceptionally straight-forward. Here are a few comments based on your issue, but this is not a full-fledged response to the entire thing as a whole (as I'm a bit time constrained right now -- but I did want to give you some response soon).

1.  As you mentioned, the [common canonical problem specification](https://github.com/exercism/problem-specifications/blob/main/exercises/dnd-character/canonical-data.json#L4-L18) gives a bit of insight on what is possible and what's not possible. If the abstract phrasing there is applied to TypeScript, it's basically saying:
    - It is impossible to test for randomness. You can only make a test that attempts to find within some reasonable probability that something is truly random. For example, a test with checks that asserts if the first 10 dice rolls are not all 6 are, however improbable, not actually testing that it's random. Getting 10 rolls of 6, assuming true randomness, is about `1.65e-8` which is small, but definitely in the space of possibility. Making the number larger makes it less probably that this test fails when something is truly random, but it doesn't _test_ for it.
    - It _is_ possible to test the _usage_ of the randomizer (in other words: is the randomizer being called when you expect it).
    - It _is_ possible to invert the tests making it more and more _probable_ that something is implemented correctly.
    - The whole thing about property based testing is more a note for implementers and not for students, and it's there because it attempts to explain how maintainers could implement this test suite in their track. 
2.  From the original issue: https://github.com/exercism/problem-specifications/issues/616 and the original PR https://github.com/exercism/problem-specifications/pull/1397 we can find some of the discussion of what you mention. Basically it was determined that we (at this point) do not at _randomness tests_ but _usage of randomness_ (for example: https://github.com/exercism/problem-specifications/pull/1397#discussion_r234162571)
3. This then means that most tests get a "best effort" approach (aka: https://github.com/exercism/problem-specifications/pull/1397#issuecomment-443751292). 
4. The `.prop === .prop` tests try to guard for `get prop() { return somethingRandom() }`, which is much more common in some other languages, but that's why it is as it is.

A few thoughts based on your observations:

1. The _order_ of the tests is sub-optimal. I will take a PR to make this better. Go for it if you feel like doing that, or we can leave this issue open is necessary.
2. The fact that none of the tests is ignored/skipped is a bug. This should be fixed. Go for it if you feel like doing that.
3. I think what may be good is mocking [`Math.rand()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Math/random) as we want people to use that anyway -- but `Crypto.getRandomValues()` technically could be used too. This would be additional tests, which I think would be a good idea here.
4. I don't understand your comment about "this exercise comes after Space Age". Can you elaborate on this, so I can help figuring out how we may solve this?

**Comment by ireddick** [comment]
After some thought:

1) I'll create a PR to initially ignore all but the first test

2) Having looked at the test names and test order in the canonical data for the exercise here:

  https://github.com/exercism/problem-specifications/blob/main/exercises/dnd-character/canonical-data.json

I see that the test order, test names, and the test with description 'each ability is only calculated once' (which tests nothing as far as I can make out) are all reflected in the tests on this track, so I'd be hesitant creating a PR to change any of these locally (I'm not sure how in-sync these things have to be).

I think at this point I'm going to say that I simply don't like this particular exercise and move on.

**Comment by SleeplessByte** [comment]
Thanks @ireddick for your issue and PR!
